File: sdk/go/vfs/serve.go
package vfs

import "io"

// Serve runs the provider request/response loop over one connection (a socket
// to the `msb` runtime). It performs the responder half of the hello handshake,
// then reads framed CBOR requests, dispatches each to `provider`, and writes the
// reply. Requests are handled one at a time, like the sibling backends, so the
// provider never sees concurrent calls on a single connection.
//
// Serve returns nil on a clean EOF (the runtime closed the channel) and the
// underlying error on any I/O failure.
func Serve(conn io.ReadWriter, provider PathFs) error {
	if _, err := readHello(conn); err != nil {
		return err
	}
	if err := writeHello(conn); err != nil {
		return err
	}

	for {
		id, payload, err := readFrame(conn)
		if err != nil {
			if err == io.EOF || err == io.ErrUnexpectedEOF {
				return nil
			}
			return err
		}

		var resp []byte
		if req, derr := decodeRequest(payload); derr != nil {
			// Mirrors the Rust server's decode_error_errno: a typed
			// limit-validation error (e.g. ENAMETOOLONG for an over-long
			// symlink target) keeps its errno; structural CBOR errors
			// report EINVAL.
			resp = respErr(errnoOr(derr, EINVAL))
		} else {
			resp = dispatch(provider, req)
		}
		if uint32(len(resp)) > maxFrameLen {
			resp = respErr(int32(EIO))
		}
		if err := writeFrame(conn, id, resp); err != nil {
			return err
		}
	}
}

File: crates/filesystem/lib/backends/vfs/rpc/serve.rs
//! Reference VFS RPC server loop: run a [`PathFs`] provider over a byte stream.
//!
//! Run this on the controlling end of a connection while the runtime serves the
//! guest via [`super::mount::unix_socket_backend`]. Like the sibling backends,
//! requests are handled one at a time, so mutating ops (e.g. rmdir's
//! emptiness check + delete) cannot race each other on the same connection.

use std::io::{self, Read, Write};
use std::sync::Arc;

use super::super::PathFs;
use super::dispatch::dispatch;
use super::mount::MountStream;
use super::protocol::{
    MAX_FRAME_LEN, VfsResponse, decode_error_errno, decode_request, read_frame, read_hello,
    to_cbor, write_frame, write_hello,
};
use crate::backends::shared::platform;

//--------------------------------------------------------------------------------------------------
// Functions
//--------------------------------------------------------------------------------------------------

/// Run the request/response loop for one virtual mount over a connected
/// AF_UNIX stream ([`MountStream`]).
pub fn serve_unix(stream: MountStream, provider: Arc<dyn PathFs>) -> io::Result<()> {
    let reader = stream.try_clone()?;
    serve(reader, stream, provider)
}

/// Like [`serve_unix`] with separate read and write halves of one connection.
///
/// Performs the responder half of the hello handshake, then reads framed CBOR
/// requests, dispatches each to `provider`, and writes replies. Returns `Ok(())`
/// on a clean EOF (the runtime closed the channel); a write failure returns the
/// error on this same thread, ending the loop cleanly.
pub fn serve<R: Read, W: Write>(
    mut reader: R,
    mut writer: W,
    provider: Arc<dyn PathFs>,
) -> io::Result<()> {
    read_hello(&mut reader)?;
    write_hello(&mut writer)?;

    loop {
        let (id, req_bytes) = match read_frame(&mut reader) {
            Ok(frame) => frame,
            Err(e) if e.kind() == io::ErrorKind::UnexpectedEof => return Ok(()),
            Err(e) => return Err(e),
        };
        let resp = match decode_request(&req_bytes) {
            Ok(req) => dispatch(provider.as_ref(), req),
            Err(err) => VfsResponse::Err(decode_error_errno(&err)),
        };
        let mut payload = to_cbor(&resp);
        if payload.len() > MAX_FRAME_LEN as usize {
            payload = to_cbor(&VfsResponse::Err(
                platform::eio()
                    .raw_os_error()
                    .unwrap_or(platform::LINUX_EIO),
            ));
        }
        write_frame(&mut writer, id, &payload)?;
    }
}
